include/small/obuf.h

```diff
@@ -223,6 +223,13 @@ obuf_alloc_cb(void *ctx, size_t size)
 	return obuf_alloc((struct obuf *) ctx, size);
 }
 
+/** Whether buffer is destroyed. */
+static inline bool
+obuf_is_destroyed(const struct obuf *buf)
```

Comment:
We have `iostream_ctx_is_initialized` and `mempool_is_initialized`. Let's introduce `obuf_is_initialized` for consistency. I'd also add `obuf_clear` that would `memset` the `obuf` object with zeroes (similarly to `iostream_ctx_clear`).

Reply:
Discussed f2f and decided to:
1. do not introduce `obuf_clear` (because of memory leaks).
2. fix `mempool_is_initialized` to return `false` after calling `mempool_destroy` in the same commit for consistency.

<details>
<summary>small diff</summary>

```diff
diff --git a/include/small/mempool.h b/include/small/mempool.h
index 2d33298..5404366 100644
--- a/include/small/mempool.h
+++ b/include/small/mempool.h
@@ -255,6 +255,10 @@ mempool_create(struct mempool *pool, struct slab_cache *cache,
 	return mempool_create_with_order(pool, cache, objsize, order);
 }
 
+/**
+ * Return true after mempool_create and false after mempool_destroy.
+ * The result of calling before mempool_create is undefined.
+ */
 static inline bool
 mempool_is_initialized(struct mempool *pool)
 {
diff --git a/include/small/obuf.h b/include/small/obuf.h
index 90f7daa..bbe7aee 100644
--- a/include/small/obuf.h
+++ b/include/small/obuf.h
@@ -115,6 +115,16 @@ struct obuf
 void
 obuf_create(struct obuf *buf, struct slab_cache *slabc, size_t start_capacity);
 
+/**
+ * Return true after obuf_create and false after obuf_destroy.
+ * The result of calling before obuf_create is undefined.
+ */
+static inline bool
+obuf_is_initialized(const struct obuf *buf)
+{
+	return buf->slabc != NULL;
+}
+
 void
 obuf_destroy(struct obuf *buf);
 
@@ -232,13 +242,6 @@ obuf_alloc_cb(void *ctx, size_t size)
 	return obuf_alloc((struct obuf *) ctx, size);
 }
 
-/** Whether buffer is destroyed. */
-static inline bool
-obuf_is_destroyed(const struct obuf *buf)
-{
-	return buf->slabc == NULL;
-}
-
 #if defined(__cplusplus)
 } /* extern "C" */
 #endif /* defined(__cplusplus) */
diff --git a/small/mempool.c b/small/mempool.c
index 03dfe5a..9a36441 100644
--- a/small/mempool.c
+++ b/small/mempool.c
@@ -183,6 +183,7 @@ mempool_destroy(struct mempool *pool)
 	rlist_foreach_entry_safe(slab, &pool->slabs.slabs,
 				 next_in_list, tmp)
 		slab_put_with_order(pool->cache, slab);
+	memset(pool, 0, sizeof(*pool));
 }
 
 void *
diff --git a/small/obuf.c b/small/obuf.c
index 4661e9f..a9b0354 100644
--- a/small/obuf.c
+++ b/small/obuf.c
@@ -95,10 +95,7 @@ obuf_destroy(struct obuf *buf)
 		struct slab *slab = slab_from_data(buf->iov[i].iov_base);
 		slab_put(buf->slabc, slab);
 	}
-#ifndef NDEBUG
-	obuf_create(buf, buf->slabc, buf->start_capacity);
-#endif
-	buf->slabc = NULL;
+	memset(buf, 0, sizeof(*buf));
 }
 
 /** Add data to the output buffer. Copies the data. */
diff --git a/test/mempool.c b/test/mempool.c
index 2982397..38fe3c6 100644
--- a/test/mempool.c
+++ b/test/mempool.c
@@ -78,18 +78,19 @@ mempool_basic()
 {
 	int i;
 
-	plan(1);
+	plan(2);
 	header();
 
 	mempool_create(&pool, &cache, objsize);
+	ok(mempool_is_initialized(&pool));
 
 	for (i = 0; i < ITERATIONS_MAX; i++) {
 		basic_alloc_streak();
 		allocating = ! allocating;
 	}
-	ok(true);
 
 	mempool_destroy(&pool);
+	ok(!mempool_is_initialized(&pool));
 
 	footer();
 	check_plan();
diff --git a/test/obuf.c b/test/obuf.c
index 3a4cc41..64b52a8 100644
--- a/test/obuf.c
+++ b/test/obuf.c
@@ -52,7 +52,7 @@ obuf_basic(struct slab_cache *slabc)
 
 	struct obuf buf;
 	obuf_create(&buf, slabc, 16320);
-	ok(!obuf_is_destroyed(&buf));
+	ok(obuf_is_initialized(&buf));
 
 	for (i = 0; i < ITERATIONS_MAX; i++) {
 		basic_alloc_streak(&buf);
@@ -61,7 +61,7 @@ obuf_basic(struct slab_cache *slabc)
 		fail_unless(obuf_size(&buf) == 0);
 	}
 	obuf_destroy(&buf);
-	ok(obuf_is_destroyed(&buf));
+	ok(!obuf_is_initialized(&buf));
 	ok(slab_cache_used(slabc) == 0);
 	slab_cache_check(slabc);
 
```

</details>

Tarantool diff:
```diff
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index f75a9da3e..c47b65826 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -1480,8 +1480,8 @@ iproto_connection_delete(struct iproto_connection *con)
         */
        ibuf_destroy(&con->ibuf[0]);
        ibuf_destroy(&con->ibuf[1]);
-       assert(obuf_is_destroyed(&con->obuf[0]));
-       assert(obuf_is_destroyed(&con->obuf[1]));
+       assert(!obuf_is_initialized(&con->obuf[0]));
+       assert(!obuf_is_initialized(&con->obuf[1]));

        assert(mh_size(con->streams) == 0);
        mh_i64ptr_delete(con->streams);
```